lib/tlog/rec.c

```diff
@@ -856,10 +867,26 @@ tlog_rec_transfer(struct tlog_errs    **perrs,
                           "Failed to set a SIGALRM signal action");
     }
 
+    /* Setup SIGCHLD signal handler */
+    sa.sa_handler = tlog_rec_sigchld_handler;
+    sigemptyset(&sa.sa_mask);
+    /* NOTE: no SA_RESTART on purpose */
+    sa.sa_flags = 0;
+    if(sigaction(SIGCHLD, &sa, NULL) == -1) {
+        grc = TLOG_GRC_ERRNO;
+        TLOG_ERRS_RAISECS(grc,
+                          "Failed to set a SIGCHLD signal action");
+    }
+
     /*
      * Transfer I/O and window changes
      */
     while (tlog_rec_exit_signum == 0) {
+        /* Exit the loop */
+        if (tlog_rec_child_exited && !read_pending) {
+            break;
+        }
```

Comment:
Wouldn't it need to be `tlog_pkt_pos_is_past(&tty_pos, &pkt)` to ensure that the tty_pos is **at**/**equal to** the same position as the packet ?

I'm having some trouble implementing this, specifically how to determine the end of process output to handle both cases:
~~~
tlog-rec-session -c ls

ssh recuser@remotehost sh -c 'echo PLATFORM; uname; echo FOUND; command -v /usr/bin/python'
~~~

The first command will eventually return errno EIO from the source, I have been successful with adding this code after the `/* Read new data */` block in tlog_rec_transfer:
~~~
        if (tlog_rec_child_exited && (grc == TLOG_GRC_FROM(errno, EIO)) && tlog_pkt_pos_is_past(&tty_pos, &pkt)) {
            break;
        }
~~~
This addresses the issue of logging and delivering complete process output from the recorded process but I am struggling with the end condition of the pipe case. Should we break when we read an zero-byte source 'output' EOF packet ? What would be the best way to ensure all output is consumed before exiting the loop? Any suggestions are appreciated.

Reply:
Yeah, `tlog_pkt_pos_is_past(&tty_pos, &pkt)` would work better, since you won't have to negate the return value :)

Yeah, hanging sounds bad, sorry about that, and the problem is probably in termination/transfer logic. If you want to rule out any JSON encoding issues, try to disable logging everything (i.e. input, output, and window changes) and see if it still behaves the same. I suspect it will. I could also recommend backing off transfer logic changes one by one and seeing what makes it hang.

I have a feeling that our main transfer loop should exit after the child has finished, and we've transferred everything from it. See if you can make that condition clearer in the code, and make sure you understand all the termination conditions together. Perhaps something will pop out at you.